main_test.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/stretchr/testify/suite"
+)
+
+var now = time.Date(time.Now().Year(), time.Now().Month(), time.Now().Day(), 0, 0, 0, 0, time.Local)
+
+type TestSuite struct {
+	suite.Suite
+
+	sourceCalServer *httptest.Server
+
+	cals map[string]*ical.Calendar
+}
+
+// newSourceCalendar creates a new source calendar with the given name, rules and modifiers
+// this is used to create a source mock source calendar which will then be used in the testing
+// of the merge process
+func (s *TestSuite) newSourceCalendar(name string, rules []config.Rule, modifiers []config.Modifier) (*ical.Calendar, config.SourceInfo, error) {
+	cal, ok := s.cals[name]
+	if ok {
+		return cal, config.SourceInfo{}, fmt.Errorf("calendar already exists")
+	}
+
+	s.cals[name] = ical.NewCalendar()
+	s.cals[name].SetName(name)
+
+	return s.cals[name], config.SourceInfo{
+		Name:      name,
+		Url:       s.sourceCalServer.URL + "/" + name + ".ics",
+		Rules:     rules,
+		Modifiers: modifiers,
+	}, nil
+}
+
+// newMockICalServer creates a new mock iCal server which will be used to serve the mock source calendars
+func (s *TestSuite) newMockICalServer() *httptest.Server {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		if !strings.EqualFold(r.URL.Path[len(r.URL.Path)-4:], icsPrefix) {
+			w.WriteHeader(http.StatusBadRequest)
+			return
+		}
+
+		calName := r.URL.Path[1 : len(r.URL.Path)-4]
+		cal, ok := s.cals[calName]
+		if !ok {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		var buf bytes.Buffer
+		cal.SerializeTo(&buf)
+		w.Write(buf.Bytes())
+	})
+	return httptest.NewServer(mux)
+}
+
+// Setup the test
+func (s *TestSuite) SetupTest() {
+	s.cals = make(map[string]*ical.Calendar)
+	s.sourceCalServer = s.newMockICalServer()
+}
+
+// Cleanup after the test
+func (s *TestSuite) TearDownSuite() {
+	s.sourceCalServer.Close()
+}
+
+// TestMergeCals tests the merging of calendars
+func (s *TestSuite) TestMergeCals() {
+	// Create a new source calendar
+	cal1, cal1SI, err := s.newSourceCalendar("cal1", []config.Rule{}, []config.Modifier{})
+	require.NoError(s.T(), err)
+	cal2, cal2SI, err := s.newSourceCalendar("cal2", []config.Rule{}, []config.Modifier{})
+	require.NoError(s.T(), err)
+
+	// Create events
+	for i := 0; i < 3; i++ {
+		n := fmt.Sprintf("event%d", i)
+		e := cal1.AddEvent(n)
+		e.SetSummary(n)
+		e.SetDescription(n)
+		e.SetStartAt(now.Add(24 * time.Hour).Add(time.Duration(6+i) * time.Hour))
+	}
+
+	// Create a merged calendar server
+	server := newTestCalServer("test", cal1SI, cal2SI)
+	defer server.Close()
+	// Fetch the calendar
+	actualCal, err := server.fetchCalendar()
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), actualCal)
+	assert.Len(s.T(), actualCal.Events(), len(cal1.Events())+len(cal2.Events()))
+
+	for _, e := range actualCal.Events() {
+		assert.NotEmpty(s.T(), e.GetProperty(ical.ComponentPropertySummary).Value)
+	}
+}
+
+func TestMain(t *testing.T) {
```

Comment:
IDE shows this as error; Wrong test signature. Tests run fine, so feel fre to ignore.

Reply:
This is the right [signature](https://pkg.go.dev/testing). What IDE are you using? Do you see a similar error on the tests that don't use the testify suite?

The reason for testify in this file is to ensure that we have room for more thorough integration tests. I'm not so familiar with the rules and modifiers and the expected output so I was hoping to set up the infrastructure and give you some simple scaffolding that you could add those tests, eventually.

Reply:
It should be
```go
func TestMain(m *testing.M) {
	log.Init("ERROR", log.NotificationService{})
	suite.Run(t, new(TestSuite))
}
```

According to GoLand. Godocs seems to agree; https://pkg.go.dev/testing#M

But testify needs *testing.T Feel free to just ignore the comment. Or change the name to not include Main if that's possible. Both are fine for me.

Reply:
Ah, that makes sense. I've never used `testing.Main` and it seems that testify won't support it https://github.com/stretchr/testify/issues/1185